src/java/org/apache/cassandra/db/compaction/TimeOrderedKeyCompactionStrategy.java

```diff
@@ -0,0 +1,950 @@
+            return Optional.empty();
+        }
+
+        List<SSTableStats> maxGarbageData = candidate.get().data;
+        OverlappingSet maxGarbageTombstones = candidate.get().tombstones;
+
+        TimeWindow dataSSTableTimeWindowMs = TimeWindow.merge(maxGarbageData.stream().map(s -> s.timeWindowMs).collect(Collectors.toList()));
+        if (logger.isDebugEnabled())
+        {
+            logger.debug("{}: candidate: {}, data_tw: {}", tableName, candidate.get(), dataSSTableTimeWindowMs);
+        }
+
+        // If there are too many files OR (data size is over the threshold and the data set is too wide),
+        // we need a sure shot way of reducing them.
+        if (maxGarbageTombstones.sstables.size() + maxGarbageData.size() > getMaxFileCountForCompaction() || dataSSTableTimeWindowMs.duration > 2 * getWindowSizeInMs())
+        {
+            if (maxGarbageTombstones.timeWindowMs.duration > 2 * getWindowSizeInMs() || maxGarbageTombstones.sstables.size() > 1)
+            {
+                logger.debug("{}: too many tombstone files", tableName);
+                return Optional.of(new SSTablesForCompaction(toSSTableReader(maxGarbageTombstones.sstables), true, true));
+            }
+
+            List<OverlappingSet> overlappingDataSets = getDistinctOverlappingSSTables(maxGarbageData);
+            long windowSzInMs = getWindowSizeInMs();
+
+            logger.debug("{}: too many data files", tableName);
+            // we are here, which means that tombstone.timerange <= 2 * window && tombstone.size() <= 1, and there could be too many data files.
+            // data files could be too wide or many small files.
+            // we would like to merge all those files that are overlapping or are in a single compaction window.
+            return overlappingDataSets.stream()
+                    // group them by compaction window
+                    .collect(Collectors.groupingBy(s -> toWindowMs(s.timeWindowMs.ts, windowSzInMs)))
+                    .values().stream()
+                    // flatten all the sstables present in all of the overlapping sets lying in this compaction window
+                    .map(v -> v.stream()
+                            .flatMap(s -> s.sstables.stream())
+                            .collect(Collectors.toList()))
+                    // get the window with the most sstables
+                    .max(Comparator.comparingInt(List::size))
+                    // limit the count of sstables to maxFileCountForCompaction
+                    .map(l -> limit(getMaxFileCountForCompaction(), l))
+                    // create a compaction task
+                    .map(l -> new SSTablesForCompaction(toSSTableReader(l), false, true));
+        }
+        else
+        {
+            List<SSTableReader> all = new ArrayList<>(toSSTableReader(maxGarbageTombstones.sstables));
+            all.addAll(toSSTableReader(maxGarbageData));
+            return Optional.of(new SSTablesForCompaction(all, false, false));
```

Comment:
And here tombstoneMerge = false means that we will delete the data but keep the corresponding tombstones, and the writer will presumably put them back out as new tombstone tables.
So we have two types of expired tombstones - either whose corresponding data is deleted, in which case they can (and should) be GCd, or whose corresponding data is not deleted, and they should not be GCd. 

I don't see how this is handled. From what I understand, we'll keep piling up these double garbage (expired and corresponding data is gone) tombstone tables, and eventually merge them due to too many files or something, but that might also result in GC of some tombstones that we want to keep, because their data is still not deleted.

Reply:
Why is splitSSTable false here? Considering that the overlap range can be huge, don't we want to split the resulting data?

Reply:
Already changed split to true.